src/ucp/proto/proto_select.c

```diff
@@ -442,49 +437,46 @@ static ucs_status_t ucp_proto_select_elem_init_thresh(
-    status = ucs_array_reserve(ucp_proto_thresh, &thresholds,
-                               ucs_array_length(&envelope));
-    if (status) {
-        goto err;
-    }
-
-    /* Allocate performance functions array */
-    status = ucs_array_reserve(ucp_proto_ranges, &perf_ranges,
-                               ucs_array_length(&envelope));
-    if (status) {
-        goto err;
-    }
-
     /* Copy the temporary thresholds list to an array inside select_elem */
     ucs_array_for_each(envelope_elem, &envelope) {
         proto_id    = envelope_elem->id;
         priv_offset = proto_init->priv_offsets[proto_id];
 
-        if (thresh_elem &&
+        if ((thresh_elem != NULL) &&
             (thresh_elem->proto_config.proto == ucp_protocols[proto_id])) {
-            thresh_elem->max_msg_length =
-                envelope_elem->range.super.max_length;
+            /* If the last element used the same protocol - extend it */
+            thresh_elem->max_msg_length
+                = envelope_elem->range.super.max_length;
         } else {
-            thresh_elem = ucs_array_append_fixed(ucp_proto_thresh, &thresholds);
-            thresh_elem->max_msg_length =
-                envelope_elem->range.super.max_length;
+            status = ucs_array_append(ucp_proto_thresh, &thresholds);
+            if (status != UCS_OK) {
+                goto err;
+            }
+
+            thresh_elem                  = ucs_array_last(&thresholds);
+            thresh_elem->max_msg_length
+                = envelope_elem->range.super.max_length;
             proto_config                 = &thresh_elem->proto_config;
             proto_config->select_param   = *proto_init->select_param;
             proto_config->ep_cfg_index   = ep_cfg_index;
             proto_config->rkey_cfg_index = rkey_cfg_index;
             proto_config->proto          = ucp_protocols[proto_id];
-            proto_config->priv           =
-                UCS_PTR_BYTE_OFFSET(select_elem->priv_buf, priv_offset);
+            proto_config->priv
+                = UCS_PTR_BYTE_OFFSET(select_elem->priv_buf, priv_offset);
         }
 
-        if (range_elem && ucp_proto_select_range_is_equal(range_elem,
+        if ((range_elem != NULL) && ucp_proto_select_range_is_equal(range_elem,
                     &envelope_elem->range)) {
-            range_elem->super.max_length =
-                envelope_elem->range.super.max_length;
+            /* If the last range had the same performance - extend it */
+            range_elem->super.max_length
+                = envelope_elem->range.super.max_length;
         } else {
-            range_elem  = ucs_array_append_fixed(ucp_proto_ranges,
-                                                 &perf_ranges);
+            status = ucs_array_append(ucp_proto_ranges, &perf_ranges);
+            if (status != UCS_OK) {
+                goto err;
+            }
+
+            range_elem  = ucs_array_last(&perf_ranges);
```

Comment:
maybe range_elem var not needed here
`*ucs_array_last(&perf_ranges) = envelope_elem->range`

Reply:
it used in 468

Reply:
right